reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,517 @@
+            )
+
+            # initialize:
+            # complex
+            map_3d_f_updated_1 = np.zeros_like(half_map_3d_f_1)
+            # complex
+            map_3d_f_updated_2 = np.zeros_like(half_map_3d_f_2)
+            # float/real
+            counts_3d_updated_1 = np.zeros_like(half_map_3d_r_1)
+            # float/real
+            counts_3d_updated_2 = np.zeros_like(half_map_3d_r_2)
+
+            for particle_idx in range(particles_f_1.shape[0]):
+                ctf_1 = ctfs_1[particle_idx]
+                ctf_2 = ctfs_2[particle_idx]
+                # Option: particle_f_1 = particles_f_1[particle_idx]
+                # Option: particle_f_2 = particles_f_2[particle_idx]
+
+                particle_f_deconv_1 = IterativeRefinement.apply_wiener_filter(
+                    particles_f_1, ctf_1, 0.01
+                )
+                particle_f_deconv_2 = IterativeRefinement.apply_wiener_filter(
+                    particles_f_2, ctf_1, 0.01
+                )
+
+                # all slices get convolved with the particle ctf
+                apply_ctf = np.vectorize(IterativeRefinement.apply_ctf_to_slice)
+
+                slices_conv_ctfs_1 = apply_ctf(slices_1, ctf_1)
+                slices_conv_ctfs_2 = apply_ctf(slices_2, ctf_2)
+
+                bayes_factors_1 = IterativeRefinement.compute_bayesian_weights(
+                    particles_f_1[particle_idx], slices_conv_ctfs_1
+                )
+                bayes_factors_2 = IterativeRefinement.compute_bayesian_weights(
+                    particles_f_2[particle_idx], slices_conv_ctfs_2
+                )
+
+                for one_slice_idx in range(bayes_factors_1.shape[0]):
+                    xyz = xyz_rotated[one_slice_idx]
+                    # if this can be vectorized, can avoid loop
+                    inserted_slice_3d_r, count_3d_r = IterativeRefinement.insert_slice(
+                        particle_f_deconv_1.real, xyz, n_pix
+                    )
+                    inserted_slice_3d_i, count_3d_i = IterativeRefinement.insert_slice(
+                        particle_f_deconv_1.imag, xyz, n_pix
+                    )
+                    map_3d_f_updated_1 += inserted_slice_3d_r + 1j * inserted_slice_3d_i
+                    counts_3d_updated_1 += count_3d_r + count_3d_i
+
+                for one_slice_idx in range(bayes_factors_2.shape[0]):
+                    xyz = xyz_rotated[one_slice_idx]
+                    inserted_slice_3d_r, count_3d_r = IterativeRefinement.insert_slice(
+                        particle_f_deconv_2.real, xyz, n_pix
+                    )
+                    inserted_slice_3d_i, count_3d_i = IterativeRefinement.insert_slice(
+                        particle_f_deconv_2.imag, xyz, n_pix
+                    )
+                    map_3d_f_updated_2 += inserted_slice_3d_r + 1j * inserted_slice_3d_i
+                    counts_3d_updated_2 += count_3d_r + count_3d_i
+
```

Comment:
There is a missing step here (sorry, I forgot to add it when I drafted this) where the Fourier 3D maps are divided out by their counts. I spoke with Fred Sigworth about how this is usually done. See the last cells of https://github.com/geoffwoollard/learn_cryoem_math/blob/master/nb/fourier_slice_2D_3D_with_trilinear.ipynb

Something like below (assuming shapes match)
```
small_num = 1
map_3d_f_norm_1 = map_3d_f_updated_1 / (small_num + counts_3d_updated_1)
```
And likewise for the `_2` half map

Reply:
When the maps are averaged together at the very end I'm not sure if it should be

```
map_3d_f_norm = 0.5*(map_3d_f_updated_1 + map_3d_f_updated_2) / (small_num + counts_3d_updated_1 + counts_3d_updated_2)
```

or 

```
map_3d_f_norm = 0.5*(map_3d_f_norm_1+map_3d_f_norm_2)
```

Reply:
As I understand it this step is needed whenever particle slices are reinserted, plus when the fsc is computed and applied it should probably use the normalized map (not completely certain on this but it doesn't seem harmful to use the normalized...) 

I also used a more wiener filter-like formula for the normalized map (a*n)/(small number + n**2)

Reply:
1. insert slices 2. normalize 3. filter with FSC. 

You want to compute the FSC once you have taken into account how the normalization should happen, as the counts among different voxels in the shell will be different, and you want to normalize each shell before you compute the correlation in each shell.

Reply:
Seems fine for the wf in the norm. 

In thinking about this, remember that the counts values of a voxel are how much was inserted at that voxel. If the count value is zero, then it  the values in the half map must be zero, because nothing was inserted there. I think this makes the issue of "blowing up the noise" less problematic.

I'll make an issue to look into this more systematically eventually. This will do for now.

Reply:
See https://github.com/compSPI/reconstructSPI/issues/25